column_buffer_go18.go

```diff
@@ -0,0 +1,994 @@
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			v := makeValueInt32(int32(*(*int16)(p)))
+			v.repetitionLevel = levels.repetitionLevel
+			v.definitionLevel = levels.definitionLevel
+			w.values = append(w.values, v)
+		}
+		_, err = w.columns[levels.columnIndex].WriteValues(w.values)
+	}
+
+	return err
+}
+
+func (w *columnBufferWriter) writeRowsInt32(rows array, size, offset uintptr, levels columnLevels) (err error) {
+	if rows.len == 0 {
+		return w.writeRowsNull(levels)
+	}
+
+	switch c := w.columns[levels.columnIndex].(type) {
+	case *int32ColumnBuffer:
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			c.values = append(c.values, *(*int32)(p))
+		}
+
+	case *uint32ColumnBuffer:
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			c.values = append(c.values, *(*uint32)(p))
+		}
+
+	default:
+		w.reset()
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			v := makeValueInt32(*(*int32)(p))
+			v.repetitionLevel = levels.repetitionLevel
+			v.definitionLevel = levels.definitionLevel
+			w.values = append(w.values, v)
+		}
+		_, err = w.columns[levels.columnIndex].WriteValues(w.values)
+	}
+
+	return err
+}
+
+func (w *columnBufferWriter) writeRowsInt64(rows array, size, offset uintptr, levels columnLevels) (err error) {
+	if rows.len == 0 {
+		return w.writeRowsNull(levels)
+	}
+
+	switch c := w.columns[levels.columnIndex].(type) {
+	case *int64ColumnBuffer:
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			c.values = append(c.values, *(*int64)(p))
+		}
+
+	case *uint64ColumnBuffer:
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			c.values = append(c.values, *(*uint64)(p))
+		}
+
+	default:
+		w.reset()
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			v := makeValueInt64(*(*int64)(p))
+			v.repetitionLevel = levels.repetitionLevel
+			v.definitionLevel = levels.definitionLevel
+			w.values = append(w.values, v)
+		}
+		_, err = w.columns[levels.columnIndex].WriteValues(w.values)
+	}
+
+	return err
+}
+
+func (w *columnBufferWriter) writeRowsInt96(rows array, size, offset uintptr, levels columnLevels) (err error) {
+	if rows.len == 0 {
+		return w.writeRowsNull(levels)
+	}
+
+	switch c := w.columns[levels.columnIndex].(type) {
+	case *int96ColumnBuffer:
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			c.values = append(c.values, *(*deprecated.Int96)(p))
+		}
+
+	default:
+		w.reset()
+		for i := 0; i < rows.len; i++ {
+			p := rows.index(i, size, offset)
+			v := makeValueInt96(*(*deprecated.Int96)(p))
+			v.repetitionLevel = levels.repetitionLevel
+			v.definitionLevel = levels.definitionLevel
+			w.values = append(w.values, v)
+		}
+		_, err = w.columns[levels.columnIndex].WriteValues(w.values)
+	}
+
+	return err
+}
+
+func (w *columnBufferWriter) writeRowsFloat32(rows array, size, offset uintptr, levels columnLevels) (err error) {
```

Comment:
Would be nice maybe to lean into code generation for some of these very repetitive functions.

Reply:
Yep.

Generics aren't in a place yet where they can yield the returns I think, we'd still need a few versions (e.g. when the type switch has more than one case), and there's a risk of seeing a measurable impact on the compute footprint.

We could use `go:generate` to render the code from a template, but this would bring quite a bit of complexity to the development process.

Typing a few functions is a one-time cost, it takes 15-20 minutes to do and the result is clear to the reader, which is an important aspect to optimize for.